Step3 - 인증을 통한 기능 구현

ssamzag opened the pull request:
안녕하세요 태경 리뷰어님!
3단계 미션 인증을 통한 기능 구현 리뷰 요청드립니다.

이번 미션은 로직적으로 크게 어려운 건 없었는데 스프링 시큐리티 관련 기능에서 좀 헤맸네요.
어노테이션만 추가하면 끝이라니.. 😢 허무하기도 하면서 스프링의 추상화에 매번 놀라네요.

무튼 이번 리뷰도 잘 부탁드리겠습니다!!

ssamzag committed:
chore: git ignore

ssamzag committed:
chore: DS_Store gitignore

ssamzag committed:
refactor: @ResetControllerAdvice 로 변경, 미사용 코드 삭제

ssamzag committed:
docs: 스탭2 피드백 정리

ssamzag committed:
docs: 3단계 미션 문서 정리

ssamzag committed:
feat&test: 3단계 미션 인증을 통한 기능 구현
- 토큰 발급 기능 인수테스트
- 내 정보 조회 인증기능 완성 및 인수테스트
- 즐겨찾기 기능 완성 및 인수테스트

ssamzag committed:
refactor: stations 파라미터 삭제하고 구현체에서 꺼내도록 리팩터링

ohtaeg commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Transactional
+    public void deleteFavorite(Long favoriteId) {
+        favoriteRepository.deleteById(favoriteId);
```
외부로부터 받는 파라미터 즐겨찾기 ID는 로그인 된 유저의 즐겨찾기가 아닐 수도 있지 않을까해요 :)
로그인 한 사람의 Id도 파라미터로 받아 로그인한 사람의 즐겨찾기인지 검증해보는 것은 어떻게 생각하시나요 :) ?

ssamzag replied:
엇 valid가 빠졌군요 ㅠㅠ.. 수정해 놓겠습니다!

ohtaeg commented on src/main/java/nextstep/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,51 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(cascade = CascadeType.PERSIST)
```
이 부분은 같이 고민해보면 좋을 것 같아요!!
회원이나 역에 대해 Cascade 옵션을 PERSIST로 설정해주신 이유가 따로 있으신가요 :) ?

저는 즐겨찾기 객체의 생성 시점은 이미 모든 역이 생성되어 있거나 회원이 존재했을 때 가능하다고 생각해서요!

물론 FavoriteService에서는 역과 회원을 조회해서 즐겨찾기 객체를 create하지만
즐겨찾기 객체만 따로 두고 봤을 때 생성되지 않은 회원이나 역이 들어온다면 CascadeType.PERSIST 설정으로 인해
생성되지 않은 회원이나 역을 즐겨찾기 생성시에 같이 생성해버릴 수도 있을 것 같아요!

그래서 저는 즐겨찾기 객체는 회원 및 역과 라이프 사이클이 다르다고 생각하였고,
회원 및 역 객체들은 즐겨찾기 생성과 무관하게 미리 생성되어 있어야 된다고 생각했어요 ㅎㅎ

물론 제 의견은 다를수도 있고, 틀릴 수도 있기에 민호님의 의견도 궁금하네요 :)
수정을 위한 리뷰가 아니오니 편하게 말씀해주시면 감사하겠습니다 🙇

ssamzag replied:
동의합니다. 즐겨찾기는 단독으로 라이프사이클을 가져야하는데 뇌없이 PERSIST를 걸었네요.
맨 마지막에 주신 코멘트처럼 id를 이용한 간접 참조로 바꾸도록 하겠습니다! 😄 
감사합니다.

ohtaeg commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,57 @@
+    @Transactional
+    public FavoriteResponse createFavorite(Long loginMemberId, FavoriteRequest favoriteRequest) {
+        Member member = memberService.findMemberOrElseThrow(loginMemberId);
+        Station sourceStation = stationService.findStationById(favoriteRequest.getSource());
+        Station targetStation = stationService.findStationById(favoriteRequest.getTarget());
+        Favorite favorite = favoriteRepository.save(new Favorite(member, sourceStation, targetStation));
```
외부로부터 받는 source 와 target이 동일한 경우가 있을수도 있다고 생각해요 :)
그럴 경우 즐겨찾기는 생성되어야할지 고민해보시면 좋을 것 같아요 😄

ssamzag replied:
음 요구사항에는 즐찾 추가 시 경로 유효성 체크가 없으니 풀어놔도 괜찮을 것 같네요!

ohtaeg commented on src/test/java/nextstep/subway/favorite/FavoriteAcceptanceTest.java:
```diff
@@ -1,8 +1,136 @@
+        accessToken = AuthAcceptanceTest.토큰_가져오기("minho@mino.com", "민호");
+    }
+
+    /**
+     * Given 지하철역, 회원등록, 로그인이 되어 있음
+     *
+     * When 즐겨찾기를 생성 요청하면
+     * Then 즐겨찾기가 생성된다
+     *
+     * When 즐겨찾기 조회를 요청하면
+     * Then 즐겨찾기 정보가 조회된다
+     *
+     * When 즐겨찾기를 삭제 요청하면
+     * Then 즐겨찾기 정보가 삭제된다
+     *
+     * When 즐겨찾기 조회를 요청하면
+     * Then 즐겨찾기 정보는 0개이다
+     */
+    @DisplayName("즐겨찾기를 관리한다")
+    @Test
+    void manageFavorite() {
```
👍

ohtaeg commented on src/test/java/nextstep/subway/favorite/FavoriteAcceptanceTest.java:
```diff
@@ -1,8 +1,136 @@
+
+    private void 즐겨찾기_정보_조회됨(ExtractableResponse<Response> findResponse) {
+        List<FavoriteResponse> favoriteResponse = getResponse(findResponse);
+        assertAll(
```
👍 👍

ohtaeg commented on src/main/java/nextstep/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,51 @@
+
+    @ManyToOne(cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "member_id")
+    private Member member;
+
+    @ManyToOne(cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "source_id")
+    private Station source;
+
+    @ManyToOne(cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "target_id")
+    private Station target;
```
해당 리뷰는 수정을 위한 리뷰가 아니오니 참고부탁드립니다 🙏 
객체간의 관계가 단방향인 경우에는 해당 객체 참조가 반드시 필요한 것인지 고민 해보신 후
필요 없다면 객체간 직접참조가 아닌 id를 이용한 간접참조도 고민해보시면 좋을 것 같아요 :)

ohtaeg commented on src/main/java/nextstep/subway/RestControllerExceptionHandler.java:
```diff
@@ -3,11 +3,11 @@
 
-@ControllerAdvice
-public class ControllerExceptionHandler {
+@RestControllerAdvice
```
👍

ohtaeg reviewed (changes requested):
3단계 구현도 잘해주셨습니다 💪 테스트도 잘해주셨어요 👍 👍 
다음 단계를 진행하기전에 변경될 부분은 없지만 한번 고민해보면 좋을 것 같은 포인트가 있어서
Request changes 하였습니다!

소소한 코멘트를 드렸는데 확인해주시면 감사하겠습니다 🙏

ohtaeg commented on src/main/java/nextstep/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,51 @@
+    private Long id;
+
+    @ManyToOne(cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "member_id")
+    private Member member;
```
해당 리뷰는 수정을 위한 리뷰가 아니오니 참고부탁드립니다 🙏 

개인적으로 회원이라는 도메인은 왠만한 도메인에서 사용되는 공통되는 도메인으로 사용되는 경우가 많다고 생각해요
이런 경우에는 회원 도메인과의 의존성을 최소화하여 회원 도메인의 복잡도를 작게 관리할 수 있을 것 같은데요!
객체간의 관계가 단방향인 경우에는 해당 객체 참조가 반드시 필요한 것인지 고민 해보신 후
필요 없다면 객체간 직접참조가 아닌 id를 이용한 간접참조도 고민해보시면 좋을 것 같아요 :)

ssamzag replied:
음 말씀하신대로 객체참조가 필요 없을 것 도메인들은 간접참조로 만드는 게 복잡도면에서 좋아 보이겠네요!
Lazy 설정을 할까하다가 즐겨찾기에서는 사용될 일이 없을 것 같으니 간접참조가 좋겠네요! 👍

ssamzag committed:
refactor: 피드백 반영
- favorite 엔티티 필드 persist 제거 및 id 간접참조로 변경
- 본인 즐겨찾기만 제거 가능하도록 로직 추가
- jpql 연습용으로 써봄

ssamzag commented:
안녕하세요 태경님!! 피드백 반영한 3단계 다시 리뷰 요청드려요.

Favorite의 필드들의 연관관계를 전부 빼면서 즐찾 조회시 Station을 다시 조회해서 맞춰줘야하는 작업이 생기더라구요. 
즐찾만 조회해서 서비스에서 처리할까하다가 JPQL을 써보고 싶어서 억지 쿼리를 넣어 보았어요. 😃 
(아니나 다를까 BNL 방식으로 조인되는데 그.. 그냥 넘어가 주십시오.. 🙏 )

그럼 주말 잘 보내십시오!!

ssamzag committed:
refactor: getStationById 리턴타입 Map 타입으로 변경

ohtaeg commented on src/main/java/nextstep/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Transactional
+    public void deleteFavoriteOnlyMine(Long memberId, Long favoriteId) {
+        Favorite favorite = findByIdOrElseThrow(favoriteId);
+        if (!favorite.createdBy(memberId)) {
+            throw new IllegalArgumentException("즐겨찾기 삭제 권한이 없습니다. memberId=" + memberId);
```
👍 👍

ohtaeg commented on src/main/java/nextstep/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,44 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private Long memberId;
+    private Long sourceStationId;
+    private Long targetStationId;
```
간접참조로 바꾸셨군요 💯

ohtaeg commented on src/main/java/nextstep/subway/station/domain/StationRepository.java:
```diff
@@ -1,10 +1,16 @@
-}
\ No newline at end of file
+
+    @Query(value = "select distinct s from Station s left join Favorite f1 on s.id = f1.sourceStationId left join Favorite f2 on s.id = f2.targetStationId where f1.memberId = :memberId or f2.memberId = :memberId")
+    Set<Station> findFavoriteStationByMemberId(@Param("memberId") Long memberId);
```
> Favorite의 필드들의 연관관계를 전부 빼면서 즐찾 조회시 Station을 다시 조회해서 맞춰줘야하는 작업이 생기더라구요.
즐찾만 조회해서 서비스에서 처리할까하다가 JPQL을 써보고 싶어서 억지 쿼리를 넣어 보았어요. 😃
(아니나 다를까 BNL 방식으로 조인되는데 그.. 그냥 넘어가 주십시오.. 🙏 )

서로 연관 관계에 대해 이해가 일치하기 때문에 학습을 위한 JPQL 사용 매우 좋다고 생각합니다 👍 👍 
저는 아무것도 못봤습니다 🙈 🙈

ohtaeg reviewed (approved):
3단계 피드백도 반영을 잘해주셨어요 💯 
이제 마지막 미션만을 남겨두고 계시네요 😄 
민호님도 주말 잘 보내시길 바라며, 마지막 미션도 화이팅입니다 🔥
